### ISSUE TITLE
typia/zod support

### ISSUE THREAD

**Issue by gotjoshua** [elysiajs/elysia#20]
any chance?

**Comment by SaltyAom** [comment]
I have this consideration I have previously with Zod/io-ts/typia.

For one, I want to put out as fast as possible while still balancing with Quality of Life, Typia introduce has 2 mature problems that make me not support Typia (at least for now).

1. It shouldn't only just check the type, the reason I pick TypeBox is that it also supports OpenAPI Schema which is important to make a standard production-grade REST API server, and the ability to specify server specs and generated documentation of fly.

> Many have asked why Elysia uses type box instead.
> 1. Typebox is built around OpenAPI which made the Swagger plug-in possible.

2. Typia requires an additional setup step and forces the developer to go TypeScript-only.

3. The cost of adding an extra union type to support End-to-End type safety is likely very high and complex, for one to make this full type-safety work, Elysia is using depends on [LocalHandler](https://github.com/elysiajs/elysia/blob/main/src/types.ts#L352), [LocalHook](https://github.com/elysiajs/elysia/blob/main/src/types.ts#L352), [ElysiaRoute](https://github.com/elysiajs/elysia/blob/main/src/types.ts#L269) [TypeRouteToEden](https://github.com/elysiajs/elysia/blob/main/src/types.ts#L299), and [CreateEden](https://github.com/elysiajs/eden/blob/main/src/types.ts#L48) which is highly a complex type consist of over 600 loc for typing only, at the moment the type inference isn't as fast as it should be but if we manage to add one more union type for Typia, it's likely to have a problem with the performance of type inference.

And in regard to performance, seems like TypeBox is experimenting with [AoT compile to improve performance](http://sinclairzx81.github.io/runtime-type-benchmarks)

## Summary
We may support Typia someday or might not support Typia at all (depending on the situation that may on fold in the future), but for now we have no interest or priority to support Typia at moment, and if we have to support 2nd validation library one day, it's likely that Zod will be our priority next.

Will leave this issues open for a week, feels free to ask more question if the answer is not cleared or if you have more questions want to ask, feels free.

**Comment by gotjoshua** [comment]
very satisfying answer, thanks @SaltyAom !

I also wasn't thrilled with the ttsc business and very much enjoyed my approach to typebox and zod.

thanks for all your work and generosity!

**Comment by murraybauer** [comment]
Whilst Typebox for OpenAPI support is a nice 'feature', these days in terms full stack javascript development (elysia is node.js) it's not always needed given the focus on server actions which use typescript types, so adding validation to exsting types using typia I feel is a nice good solution.

I don't really see that big a difference DX wise from Typebox to Zod.

**Comment by mildronize** [comment]
@SaltyAom  I've found the interesting repo: https://github.com/decs/typeschema for Universal adapter for TypeScript schema validation.

They support various type of data validation. It might be help.

**Comment by kravetsone** [comment]
> I don't really see that big a difference DX wise from Typebox to Zod.

Unfortunately, there is a difference in DX... The TypeBox does nothing more than the JSON SCHEMA allows itself to do
Zod has excellent chainable validators/transformers, which greatly simplifies development
![image](https://github.com/elysiajs/elysia/assets/57632712/df42d40f-70f2-4dfe-9818-393f7c2251c4)

By the way, there is an interesting alternative to Zod - Valibot ([Site](https://valibot.dev/), [Github](https://github.com/fabian-hiller/valibot)) 




> FWIW Zod type-checking performance has historically been pretty poor as projects increase in complexity. However, I have not checked on a real-world project lately. Most recent deep dive I've found -> https://dev.to/nicklucas/typescript-runtime-validators-and-dx-a-type-checking-performance-analysis-of-zodsuperstructyuptypebox-5416

It's impressive! But we need to make sure that we don't lose anything from integrating this
and what if elysia has its own contribution to the type system
```ts
t.Object({
    avatar: t.File({
        type: "image",
    }),
}),
```
I'm not ready to give it up! It's too convenient

**Comment by em-jones** [comment]
> I don't really see that big a difference DX wise from Typebox to Zod.

Two things that "jump out" at me are:
- branded types

```typescript
const NonEmptyString = z.string().min(1).brand("NonEmptyString");
type NonEmptyString = z.infer<typeof NonEmptyString>;

const greet = (s: NonEmptyString) => z.string().min(1);
```
This sort of approach is handy for protecting your system from invalid data at the type level.

- ubiquity... Not to say that this _alone_ should be a deciding factor, but a tool's popularity *can* be a benefit

**Comment by medv** [comment]
Saw the trpc plugin and decided to give elysia a go, sure enough spotted the "compile" import in the example only after hitting a type error passing zod schemas to the procedures. Definitely a non-starter, would have been an instant switch from fastify otherwise. Even for a greenfield project like this, many folks already have their big bag of zod tools and patterns they would like to slingshot a new project with. For example https://github.com/chrishoermann/zod-prisma-types is alone responsible for dozens if not hundreds of hours saved during prototyping phase. The switch over to a proprietary type is a tall ask and definitely stunts this project.

Sorry, I know how this probably reads, but I've a good feeling that many people are not bothering to say anything and instead have just gone to use something else.

**Comment by em-jones** [comment]
@medv 's sentiment:
>  Definitely a non-starter, would have been an instant switch from fastify otherwise.

This is an anecdotal scenario of why ubiquity can be meaningful. 

> (not to mention react-hook-form and its zodResolver on the other side). 
To their point... what's stopping this from being an extensible interface? (besides the obnoxiousness that might come from some additional generic typing)

**Comment by fernandortec** [comment]
https://github.com/turkerdev/fastify-type-provider-zod

Can this be of any help?
This package can create swagger docs using all Zod schemas and Fastify, would an implementation like this solve the issue of generating the open API schema?

**Comment by DoK6n** [comment]
[https://typia.io/docs/setup/#unplugin-typia](https://typia.io/docs/setup/#unplugin-typia)
[https://github.com/ryoppippi/unplugin-typia/tree/main/examples/bun-build](https://github.com/ryoppippi/unplugin-typia/tree/main/examples/bun-build)

Typia can now be used in Next.js and Bun environments.
While supporting extra union types for End-to-End type safety is challenging, validation can be done before the request.

**Comment by SaltyAom** [comment]
Closing as not planned at the moment, however feels free to continue to discussion or open a new discussion in the discussion tab.

Thanks.

**Comment by jasperdunn** [comment]
@SaltyAom the reason I'm keen for Zod support is being able to do cross-field validation.

Is there any possible way to do the following in Elysia atm?

```ts
t.Object({
  max: t.Number(),
  min: t.Number({ minimum: 0 }),
})
```
validate that minimum is less than maximum and that maximum is less than minimum at runtime?

**Comment by jasperdunn** [comment]
Ahh I just saw [beforeHandle](https://elysiajs.com/life-cycle/before-handle) in the docs 👍🏼

**Comment by PandaWorker** [comment]
Here it remains only to use the type box or write your own validator/serializer, which will be completely covered by types and JsonSchema last draft(2020-12) + OpenAPI 3.1.0 support.

**Comment by ismailkarsli** [comment]
In a framework that works like magic, it's sad to write validations manually. It's even sadder that by far the most upvoted issue is marked as wontfix.

Hono has Typia support but OpenAPI support is not sufficient. I don't know what to do [🤷‍♂️](https://emojipedia.org/man-shrugging)

**Comment by ismailkarsli** [comment]
> It shouldn't only just check the type, the reason I pick TypeBox is that it also supports OpenAPI Schema which is important to make a standard production-grade REST API server, and the ability to specify server specs and generate documentation of fly.

I can't say anything about other reasons, you're probably right, but I should specify that Typia has a JSON Schema generation feature: https://typia.io/docs/json/schema/

It can even derive descriptions and rules from interfaces directly.

**Comment by jacksteamdev** [comment]
The creator of Zod has proposed a [Standard Schema API](https://github.com/standard-schema/standard-schema) for libraries to support instead of a single validation library. 
Currently, Arktype, Valibot, and Zod support the Standard Schema, but nothing is stopping others from adopting it.

List of current libraries consuming it: https://github.com/standard-schema/standard-schema?tab=readme-ov-file#ecosystem

**Comment by kravetsone** [comment]
> The creator of Zod has proposed a [Standard Schema API](https://github.com/standard-schema/standard-schema) for libraries to support instead of a single validation library. 
> Currently, Arktype, Valibot, and Zod support the Standard Schema, but nothing is stopping others from adopting it.
> 
> List of current libraries consuming it: https://github.com/standard-schema/standard-schema?tab=readme-ov-file#ecosystem

Seems interesting

**Comment by jasperdunn** [comment]
Issues like [this one](https://github.com/sinclairzx81/typebox/discussions/943) make me feel like TypeBox is just not at all designed for UX, instead it focuses solely on performance whilst checking and parsing data.

Most use cases for me and anyone working on web apps will be form validation with custom error messages and i18n. I really think that Zod should be considered for a project like Elysia, since it's mature and well designed for custom error messages and form validation.

Looks like someone worked on a branch to try this out https://github.com/Savolus/elysia-with-zod

**Comment by kravetsone** [comment]
> Issues like [this one](https://github.com/sinclairzx81/typebox/discussions/943) make me feel like TypeBox is just not at all designed for UX, instead it focuses solely on performance whilst checking and parsing data.
> 
> Most use cases for me and anyone working on web apps will be form validation with custom error messages and i18n. I really think that Zod should be considered for a project like Elysia, since it's mature and well designed for custom error messages and form validation.
> 
> Looks like someone worked on a branch to try this out https://github.com/Savolus/elysia-with-zod

Typebox designed for perfomance and JSON Schema Builder

I guess our community can create new one validator that aims DX and perfomance

a lot of things of DX can be JIT'ed to improve perfomance

**Comment by rjzheng** [comment]
need this!

**Comment by enriquecaballero-agx** [comment]
I really tried translating my Zod schemas to TypeBox because of the general performance benefits Elysia offers. But man, I couldn’t pull it off.

I thought I’d install the main TypeBox module and start from there. Then I realized I couldn’t use that official TypeBox module but had to use Elysia.t. Is that correct?

I can’t justify making my entire project dependent on a single web framework, especially a brand new one. The coupling is too much, which increases the risk.

While coupling Elysia to TypeBox is theoretically sound with legitimate reasoning, it’s not very pragmatic; I hope a Zod to Elysia.t adapter becomes available at some point. Also, why wrap TypeBox? Why not have an adapter that makes existing TypeBox schemas Elysia compatible?

I love the work done here; there’s lots of great stuff. I’ll keep an eye on how it progresses. Definitely in the future, starting fresh, I’ll give it a try, especially since I’m using Bun for all of my Node projects nowadays.

**Comment by enriquecaballero-agx** [comment]
I really tried translating my Zod schemas to TypeBox because of the general performance benefits Elysia offers. But man, I couldn’t pull it off.

I thought I’d install the main TypeBox module and start from there. Then I realized I couldn’t use that official TypeBox module but had to use Elysia.t. Is that correct?

I can’t justify making my entire project dependent on a single web framework, especially a brand new one. The coupling is too much, which increases the risk.

While coupling Elysia to TypeBox is theoretically sound with legitimate reasoning, it’s not very pragmatic; I hope a Zod to Elysia.t adapter becomes available at some point. Also, why wrap TypeBox? Why not have an adapter that makes existing TypeBox schemas Elysia compatible?

I love the work done here; there’s lots of great stuff. I’ll keep an eye on how it progresses. Definitely in the future, starting fresh, I’ll give it a try, especially since I’m using Bun for all of my Node projects nowadays.

**Comment by kravetsone** [comment]
> I really tried translating my Zod schemas to TypeBox because of the general performance benefits Elysia offers. But man, I couldn’t pull it off.
> 
> I thought I’d install the main TypeBox module and start from there. Then I realized I couldn’t use that official TypeBox module but had to use Elysia.t. Is that correct?
> 
> I can’t justify making my entire project dependent on a single web framework, especially a brand new one. The coupling is too much, which increases the risk.
> 
> While coupling Elysia to TypeBox is theoretically sound with legitimate reasoning, it’s not very pragmatic; I hope a Zod to Elysia.t adapter becomes available at some point. Also, why wrap TypeBox? Why not have an adapter that makes existing TypeBox schemas Elysia compatible?
> 
> I love the work done here; there’s lots of great stuff. I’ll keep an eye on how it progresses. Definitely in the future, starting fresh, I’ll give it a try, especially since I’m using Bun for all of my Node projects nowadays.

`Elysia.t` is just typebox with some extends (like t.File)

**Comment by enriquecaballero** [comment]
> > I really tried translating my Zod schemas to TypeBox because of the general performance benefits Elysia offers. But man, I couldn’t pull it off.
> > I thought I’d install the main TypeBox module and start from there. Then I realized I couldn’t use that official TypeBox module but had to use Elysia.t. Is that correct?
> > I can’t justify making my entire project dependent on a single web framework, especially a brand new one. The coupling is too much, which increases the risk.
> > While coupling Elysia to TypeBox is theoretically sound with legitimate reasoning, it’s not very pragmatic; I hope a Zod to Elysia.t adapter becomes available at some point. Also, why wrap TypeBox? Why not have an adapter that makes existing TypeBox schemas Elysia compatible?
> > I love the work done here; there’s lots of great stuff. I’ll keep an eye on how it progresses. Definitely in the future, starting fresh, I’ll give it a try, especially since I’m using Bun for all of my Node projects nowadays.
> 
> `Elysia.t` is just typebox with some extends (like t.File)

Try passing in a raw TypeBox schema not built with t.

**Comment by kravetsone** [comment]
> > > I really tried translating my Zod schemas to TypeBox because of the general performance benefits Elysia offers. But man, I couldn’t pull it off.
> > > I thought I’d install the main TypeBox module and start from there. Then I realized I couldn’t use that official TypeBox module but had to use Elysia.t. Is that correct?
> > > I can’t justify making my entire project dependent on a single web framework, especially a brand new one. The coupling is too much, which increases the risk.
> > > While coupling Elysia to TypeBox is theoretically sound with legitimate reasoning, it’s not very pragmatic; I hope a Zod to Elysia.t adapter becomes available at some point. Also, why wrap TypeBox? Why not have an adapter that makes existing TypeBox schemas Elysia compatible?
> > > I love the work done here; there’s lots of great stuff. I’ll keep an eye on how it progresses. Definitely in the future, starting fresh, I’ll give it a try, especially since I’m using Bun for all of my Node projects nowadays.
> > 
> > 
> > `Elysia.t` is just typebox with some extends (like t.File)
> 
> Try passing in a raw TypeBox schema not built with t.

it is because elysia uses -1 major of typebox)

**Comment by enriquecaballero** [comment]
> > > > I really tried translating my Zod schemas to TypeBox because of the general performance benefits Elysia offers. But man, I couldn’t pull it off.
> > > > I thought I’d install the main TypeBox module and start from there. Then I realized I couldn’t use that official TypeBox module but had to use Elysia.t. Is that correct?
> > > > I can’t justify making my entire project dependent on a single web framework, especially a brand new one. The coupling is too much, which increases the risk.
> > > > While coupling Elysia to TypeBox is theoretically sound with legitimate reasoning, it’s not very pragmatic; I hope a Zod to Elysia.t adapter becomes available at some point. Also, why wrap TypeBox? Why not have an adapter that makes existing TypeBox schemas Elysia compatible?
> > > > I love the work done here; there’s lots of great stuff. I’ll keep an eye on how it progresses. Definitely in the future, starting fresh, I’ll give it a try, especially since I’m using Bun for all of my Node projects nowadays.
> > > 
> > > 
> > > `Elysia.t` is just typebox with some extends (like t.File)
> > 
> > 
> > Try passing in a raw TypeBox schema not built with t.
> 
> it is because elysia uses -1 major of typebox)

I'm...not following? My point was that I would need to `import { t } from "elysia"` everywhere in my application in order to build Elysia-compatible TypeBox schemas. If this is incorrect, then please show me where I'm mistaken. I _want_ to be mistaken. TypeBox is great in it of itself because it's generic. Can work with any web framework―I can make it work with Express, Fastify, Hono, etc. Elysia.t is exclusive to Elysia. I would be coupling my entire codebase to Elysia.

**Comment by ismailkarsli** [comment]
Just import from “@sinclair/typebox” package without installing anything. It should use Elysia version of Typebox. @enriquecaballero

**Comment by enriquecaballero** [comment]
> Just import from “@sinclair/typebox” package without installing anything. It should use Elysia version of Typebox. @enriquecaballero

I tried that...

**Comment by ismailkarsli** [comment]
> > Just import from “@sinclair/typebox” package without installing anything. It should use Elysia version of Typebox. @enriquecaballero
> 
> I tried that...

Don’t install @sinclair/typebox package too. I’m using this way and works.

**Comment by enriquecaballero** [comment]
> > > Just import from “@sinclair/typebox” package without installing anything. It should use Elysia version of Typebox. @enriquecaballero
> > 
> > 
> > I tried that...
> 
> Don’t install @sinclair/typebox package too. I’m using this way and works.

I would rather use @sinclair/typebox as _the_ dependency in my code and then use the schemas defined with @sinclair/typebox with Elysia. NOT the other way around.

**Comment by kravetsone** [comment]
> > > > Just import from “@sinclair/typebox” package without installing anything. It should use Elysia version of Typebox. @enriquecaballero
> > > 
> > > 
> > > I tried that...
> > 
> > Don’t install @sinclair/typebox package too. I’m using this way and works.
> 
> I would rather use @sinclair/typebox as _the_ dependency in my code and then use the schemas defined with @sinclair/typebox with Elysia. NOT the other way around.

Override typebox version in elysia

**Comment by enriquecaballero-agx** [comment]
> > > > > Just import from “@sinclair/typebox” package without installing anything. It should use Elysia version of Typebox. @enriquecaballero
> > > > 
> > > > 
> > > > I tried that...
> > > 
> > > 
> > > Don’t install @sinclair/typebox package too. I’m using this way and works.
> > 
> > 
> > I would rather use @sinclair/typebox as _the_ dependency in my code and then use the schemas defined with @sinclair/typebox with Elysia. NOT the other way around.
> 
> Override typebox version in elysia

… Can you elaborate be more specific? As I said in my original comment, I tried to make this work and I tried various things. It’s not so trivial.

**Comment by sinclairzx81** [comment]
@enriquecaballero-agx @enriquecaballero @kravetsone @jasperdunn @rjzheng 

> I hope a Zod to Elysia.t adapter becomes available at some point.

A new TypeBox project is now available, allowing you to integrate Zod and Valibot into Elysia. This project transforms Zod and Valibot types into TypeBox-compatible schematics (i.e., JSON Schema), enabling acceleration by the TypeBox compiler (matching TypeBox's speed) and compatibility with OpenAPI/Swagger.

Project: https://github.com/sinclairzx81/typebox-remix

---

### Example

The following creates an Elysia route and configures parameters using TypeBox, Zod and Valibot. Type inference is preserved.

[TypeScript Link Here](https://www.typescriptlang.org/play/?moduleResolution=99&module=199#code/JYWwDg9gTgLgBAbzgIQgDzgXzgMyhEOAcgAEBnYAOwGMAbAQ2CgHoYBPMAUwCN0BaKJxDA0RAFBjQkWIjgBRWmwr0ANHHjY8BYp0XLxU6PABeufISLGIAEwPgjcAG5ntRR-VrBeMcWMqcAd3k9YHoACgBKMTg4ADoAc04YMKJmQWFRNTCkMHooehAyNQBHAFdOKDY1Xms2LAi4AF4APkRomLhqCEoyeCQ0NSq4U2xGuFz8wrhmZhdCABUOTlQ0dpiunr64VThuNWosJrgyirrp2a1CAC0bNc7u3tki9TVSw7Gas-O5uAA1Dy8EBg7UwagQdwmBTIAC51LEAPLcABWnGoyXBHQ6aFhMFiAGUYFAqPFIio7jE2DjYgA5UogbgVUnk4ZU1AQWiceiUSJ3TARMl3E6VWErMLGWIQZGo9HM+iw8W9ImUEn85nceWxSh0hlQJmY9Ya3jszncqKYvmqu6fEXoMKOCVStHZZkwpyxRXEvX6mCw+1a+mM1X6uClX2xI0crk880RIP1dqxTy9TjcgDMAAZMxEgA)

```typescript
import { Box } from '@sinclair/typebox-remix'

import { Elysia, t } from 'elysia'
import z from 'zod'
import v from 'valibot'

new Elysia()
  .get('/remix', ({ params, query, body }) => {
    const { x, y, z } = params // from TypeBox
    const { a, b, c } = query  // from Zod
    const { s, t, u } = body   // from Valibot
  }, {
    params: t.Object({
      x: t.String(),
      y: t.Number(),
      z: t.Boolean()
    }),

    query: Box(z.object({
      a: z.string(),
      b: z.number(),
      c: z.boolean()
    })),

    body: Box(v.object({
      s: v.string(),
      t: v.number(),
      u: v.boolean()
    })),
  })
  .listen(3000)
```

---

This project is ongoing, with work continuing through 2025. I anticipate additional updates will be required when Zod 4 is released, and there are considerations to include other libraries like EffectTS in the future. If you use this project, feel free to submit issues or provide general feedback.

Hope this helps
S

**Comment by kravetsone** [comment]
> @enriquecaballero-agx @enriquecaballero @kravetsone @jasperdunn @rjzheng
> 
> > I hope a Zod to Elysia.t adapter becomes available at some point.
> 
> A new TypeBox project is now available, allowing you to integrate Zod and Valibot into Elysia. This project transforms Zod and Valibot types into TypeBox-compatible schematics (i.e., JSON Schema), enabling acceleration by the TypeBox compiler (matching TypeBox's speed) and compatibility with OpenAPI/Swagger.
> 
> Project: https://github.com/sinclairzx81/typebox-remix
> 
> ### Example
> The following creates an Elysia route and configures parameters using TypeBox, Zod and Valibot. Type inference is preserved.
> 
> [TypeScript Link Here](https://www.typescriptlang.org/play/?moduleResolution=99&module=199#code/JYWwDg9gTgLgBAbzgIQgDzgXzgMyhEOAcgAEBnYAOwGMAbAQ2CgHoYBPMAUwCN0BaKJxDA0RAFBjQkWIjgBRWmwr0ANHHjY8BYp0XLxU6PABeufISLGIAEwPgjcAG5ntRR-VrBeMcWMqcAd3k9YHoACgBKMTg4ADoAc04YMKJmQWFRNTCkMHooehAyNQBHAFdOKDY1Xms2LAi4AF4APkRomLhqCEoyeCQ0NSq4U2xGuFz8wrhmZhdCABUOTlQ0dpiunr64VThuNWosJrgyirrp2a1CAC0bNc7u3tki9TVSw7Gas-O5uAA1Dy8EBg7UwagQdwmBTIAC51LEAPLcABWnGoyXBHQ6aFhMFiAGUYFAqPFIio7jE2DjYgA5UogbgVUnk4ZU1AQWiceiUSJ3TARMl3E6VWErMLGWIQZGo9HM+iw8W9ImUEn85nceWxSh0hlQJmY9Ya3jszncqKYvmqu6fEXoMKOCVStHZZkwpyxRXEvX6mCw+1a+mM1X6uClX2xI0crk880RIP1dqxTy9TjcgDMAAZMxEgA)
> 
> ```ts
> import { Box } from '@sinclair/typebox-remix'
> 
> import { Elysia, t } from 'elysia'
> import z from 'zod'
> import v from 'valibot'
> 
> new Elysia()
>   .get('/remix', ({ params, query, body }) => {
>     const { x, y, z } = params // from TypeBox
>     const { a, b, c } = query  // from Zod
>     const { s, t, u } = body   // from Valibot
>   }, {
>     params: t.Object({
>       x: t.String(),
>       y: t.Number(),
>       z: t.Boolean()
>     }),
> 
>     query: Box(z.object({
>       a: z.string(),
>       b: z.number(),
>       c: z.boolean()
>     })),
> 
>     body: Box(v.object({
>       s: v.string(),
>       t: v.number(),
>       u: v.boolean()
>     })),
>   })
>   .listen(3000)
> ```
> 
> This project is ongoing, with work continuing through 2025. I anticipate additional updates will be required when Zod 4 is released, and there are considerations to include other libraries like EffectTS in the future. If you use this project, feel free to submit issues or provide general feedback.
> 
> Hope this helps S

Fine! But i guess remix not so good name because react-remix))

also did u want to work on new validation library? Like Typebox but with more DX and JITed

Like Zod but really faster (for example compile string().maxLength(8, "error").trim() to a function)

JSON schema needed but can be generated by remappers from internal

**Comment by ismailkarsli** [comment]
> > @sinclairzx81 Do you think there will be a chance to add Typia support in the future?
> 
> typia support is hard
> 
> because it AOT with builder plugin

I’m currently converting [schema()](https://typia.io/docs/llm/schema/) function output (which is JSON schema) to Typebox schema. But the lack of first-party support bothers me.

**Comment by sinclairzx81** [comment]
@kravetsone Hi,

> Fine! But i guess remix not so good name because react-remix))

I agree. Remix was the working title used during development, but I guess it's not as descriptive as it could be. I have republished under a the more formal name `@sinclair/typebox-adapter`. 

github: https://github.com/sinclairzx81/typebox-adapter
npm: https://www.npmjs.com/package/@sinclair/typebox-adapter

Thanks for the feedback :)

**Comment by kravetsone** [comment]
> @kravetsone Hi,
> 
> > Fine! But i guess remix not so good name because react-remix))
> 
> I agree. Remix was the working title used during development, but I guess it's not as descriptive as it could be. I have republished under a the more formal name `@sinclair/typebox-adapter`. 
> 
> github: https://github.com/sinclairzx81/typebox-adapter
> npm: https://www.npmjs.com/package/@sinclair/typebox-adapter
> 
> Thanks for the feedback :)

Great!

**Comment by aliabdy** [comment]
Hi, I think it would be great if Elysia adopted this standard. It can solve the issue and allow users to pick any validation they want.
[](https://standardschema.dev/)

**Comment by aliabdy** [comment]
Hi, I think it would be great if Elysia adopted this standard. It can solve the issue and allow users to pick any validation they want.
[https://standardschema.dev/](url)

**Comment by markH5** [comment]
> Hi, I think it would be great if Elysia adopted this standard. It can solve the issue and allow users to pick any validation they want. [https://standardschema.dev/](url)

your link  is incorrect, you can use the following syntax
```md
<https://standardschema.dev/> or [some text](https://standardschema.dev/)

not [https://standardschema.dev/](url)
```

**Comment by ismailkarsli** [comment]
> Hi, I think it would be great if Elysia adopted this standard. It can solve the issue and allow users to pick any validation they want. [https://standardschema.dev/](url)

I think Standard Schema is not valid for Elysia. Elysia uses Typebox for both validation and JSON Schema generation. Standard Schema offers validation but does not have an interface for JSON Schema.

But I hope I am wrong. I would like to use Elysia with Typia.

**Comment by JagritGumber** [comment]
Type inference with eden is actually an issue, when I make too complicated paths like `api/v1/{some}/{some}` then eden treaty does stop working at the final route, but edenfetch does work, Idk why is that

The reason for not integrating typia or any other library sounds quite convincing with the evidence I have seen I am guessing eden fetch works because it is using string keys for routes actually causes less nested types because of which it works. 

Adding another Union would increase the complexity of Eden Treaty by a lot. It might just stop working for me atleast all together

Though I would love to see a zod integration, It has been my go to choice for quite a while

**Comment by carere** [comment]
The only reason I don't use Elysia is because it force me to use Typebox, and I don't like the API of Typebox.
Actually, i'm using oRPC, and this lib does support OpenAPI spec generation and we can use any validation library which support the Standard Schema Validator. Can't wait to see this coming in ElysiaJs 😄

**Comment by SaltyAom** [comment]
#1381 will support Standard Schema on Elysia 1.4

You can try it out now on `1.4.0-exp.5`

<img src=https://github.com/user-attachments/assets/d0c962e8-5c74-47c4-80aa-675b6d9d34e0 alt=nihamazing height=160 />

**Comment by boeto** [comment]
@SaltyAom I have already tried it, great update! But they are not yet properly displayed in '@elysiajs/openapi', any plans for this? [Request support for Standard Schema](https://github.com/elysiajs/elysia-openapi/issues/245)

**Comment by SaltyAom** [comment]
@boeto we have just added support Standard Schema support for OpenAPI via external library in `@elysiajs/openapi` `1.4.0-exp.0`

It would be nice for standard schema to support JSON Schema out of the box, but as for now we have some workaround for that

**Comment by boeto** [comment]
Hearing this news excites me, and I can now migrate the project's validation methods to Zod.
